comp/core/autodiscovery/scheduler/configstate.go

```diff
@@ -0,0 +1,120 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package scheduler
+
+import (
+	"sync"
+
+	"github.com/DataDog/datadog-agent/comp/core/autodiscovery/integration"
+)
+
+// ConfigState represents the state of the config: scheduled or unscheduled
+type ConfigState int8
+
+// TaskStatus represents the status of the schedule task: success, failed or config not found
+type TaskStatus int8
+
+// Digest is the unique identifier of the config
+type Digest string
+
+const (
+	// Scheduled config should be scheduled or is scheduled
+	Scheduled ConfigState = 1 << iota
+	// Unscheduled config is unscheduled or should be unscheduled
+	Unscheduled
+)
+
+// ConfigStateData contains
+// desiredState which is the eventually desired state of the config (to be scheduled or unscheduled)
+type ConfigStateData struct {
+	desiredState ConfigState
+	config       *integration.Config
+}
+
+// copy returns a copy of ConfigStateData
+func (c ConfigStateData) copy() ConfigStateData {
+	var configCopy *integration.Config
+	originalConfig := *c.config
+	configCopy = &originalConfig
+	return ConfigStateData{
+		desiredState: c.desiredState,
+		config:       configCopy,
+	}
+}
+
+// ConfigStateStore is in charge of storing
+// 1. the scheduled integration.Config
+// 2. desired state of the config to be scheduled/unscheduled compared to existing desiredState + currentState:
+// for example, a new config has been called
+// with [schedule, unschedule, schedule], the config will be scheduled, but only one schedule should be executed
+type ConfigStateStore struct {
+	configsLock sync.Mutex
+
+	// Events to Controller would update immediately configStatusMap
+	configStateMap map[Digest]ConfigStateData
```

Comment:
How is the map cleaned? I would expect an item to be deleted when the desired state is `Unscheduled` and that it was successfully unscheduled.

Note that in this flow, you have a race by nature:
```
GetConfigState() returns `desiredState: Unschedule`
-> An AD providers calls `ApplyChanges`, update state to `desiredState: Schedule`
-> After `Unschedule` is called, you call `delete` on a `desiredState: Schedule`
```

Reply:
It won't be cleaned, otherwise it will have race condition. Only applychanges can change configStateMap in current flow. However, we can remove "config       *integration.Config" data once unschedule is sucessful